doc/designs/quic-design/quic-event-loop.md

```diff
@@ -0,0 +1,160 @@
+# Design Problem: Event Loop
+
+Applications need to be able to react to events on:
+
+* Sockets
+* Connections
+* Streams
+* Timers
+
+Only the OpenSSL library can supply the events for Connections and Streams.
+Events on sockets and timers ultimately come from a system source - the
+application may want to receive the event directly from the system, or from
+OpenSSL.
+
+For the MVP a blocking mode on top of one of the options below is likely to be
+required because that’s the default mode for `s_client`.
+
+## Candidate Solution: Application Selects
+
+In this approach the application "selects" (or whatever alternative function
+the applications wishes to use) on the socket and asks OpenSSL to process
+incoming packets when they arrive and tell it about any connection/stream
+events. The application implements the event loop.  Other calls have to be made
+to deal with each of the event possibilities. For timers, OpenSSL supplies a
+function to return how long until the next timer event so that the application
+knows how long to wait.
+
+Pros:
+
+* The application has full control over the event loop and can implement it in
+whichever way most makes sense to the application
+* Closest solution to the current API for SSL/TLS/DTLS
+* Enables the application to handle interaction with the network layer - which
+could be abstracted from OpenSSL via BIOs etc
+
+Cons:
+
+* Requires a lot of "boiler plate" code to get anything working
+* The application has to handle timers itself (unlike applications written for
+SSL/TLS)
+* Not so easy to write the simplest application that want it to "just work"
+
+![alt_text](images/image2.png "Application Selects Sequence Diagram")
```

Comment:
Looking at this diagram I am missing the option to have the application to call the `readmsg()` on its own and just pass the read data into the OpenSSL. So instead of `OSSL_process_packets(fds)` it would be `OSSL_process_packets(data, len)`. Perhaps with some way to indicate the peer IP address where the data was received from.

Reply:
For TLS/DTLS we provide this capability via a BIO. So the application specifies what BIO to use. Libssl then calls BIO_read(), which then does the low level network call. I would assume that we would do something similar. This could probably apply to most (all?) of these options. Do we need to show this detail?

Reply:
> Do we need to show this detail?

Perhaps we do not need to present this in the pictures. But it would be nice to mention it somewhere so we avoid further related questions.